File: lock-free-stack-using-kotlin/src/main/java/c/hyeoni/lockfreestackusingjava/ClassicStack.java
package c.hyeoni.lockfreestackusingjava;

public class ClassicStack<T> {

    private StackNode<T> headNode;
    private int noOfOperations;

    public synchronized int getNoOfOperations() {
        return noOfOperations;
    }

    public synchronized void push(T number) {
        StackNode<T> newNode = new StackNode<>(number);
        newNode.next = headNode;
        headNode = newNode;
        noOfOperations++;
    }

    public synchronized T pop() {
        if (headNode == null) {
            return null;
        } else {
            T val = headNode.getValue();
            headNode = headNode.next;
            noOfOperations++;
            return val;
        }
    }

    private static class StackNode<T> {

        T value;
        StackNode<T> next;

        public StackNode(final T value) {
            this.value = value;
        }

        public T getValue() {
            return value;
        }
    }
}

File: lock-free-stack-using-kotlin/src/main/kotlin/c/hyeoni/lockfreestackusingkotlin/ClassicStack.kt
package c.hyeoni.lockfreestackusingkotlin


class ClassicStack<T> {
    private var headNode: StackNode<T>? = null

    @get:Synchronized
    var noOfOperations = 0
        private set

    @Synchronized
    fun push(number: T) {
        val newNode = StackNode(number)
        newNode.next = headNode
        headNode = newNode
        noOfOperations++
    }

    @Synchronized
    fun pop(): T? {
        return if (headNode == null) {
            null
        } else {
            val value = headNode!!.value
            headNode = headNode!!.next
            noOfOperations++
            value
        }
    }

    private class StackNode<T>(var value: T) {
        var next: StackNode<T>? = null
    }
}

